lib/src/metta/runner/modules/catalog.rs

```diff
@@ -0,0 +1,568 @@
+            match mod_name {
+                Some(mod_name) => Some(Box::new(SingleFileModule::new(mod_name, path))),
+                None => {
+                    let mod_name = path.file_stem().unwrap().to_str().unwrap();
+                    Some(Box::new(SingleFileModule::new(mod_name, path)))
+                }
+            }
+        } else {
+            None
+        }
+    }
+}
+
+/// An object to identify and load a MeTTa module implemented as a directory
+#[derive(Debug)]
+pub struct DirModuleFmt;
+
+impl FsModuleFormat for DirModuleFmt {
+    fn paths_for_name(&self, parent_dir: &Path, mod_name: &str) -> Vec<PathBuf> {
+        let path = parent_dir.join(mod_name);
+        vec![path]
+    }
+    fn try_path(&self, path: &Path, mod_name: Option<&str>) -> Option<Box<dyn ModuleLoader>> {
+        if path.is_dir() {
+            let mod_matta_path = path.join("module.metta");
+            if mod_matta_path.exists() {
+                return match mod_name {
+                    Some(mod_name) => Some(Box::new(DirModule::new(mod_name, path))),
+                    None => {
+                        let mod_name = path.file_stem().unwrap().to_str().unwrap();
+                        Some(Box::new(DirModule::new(mod_name, path)))
+                    }
+                };
+            }
+        }
+        None
+    }
+}
+
+/// Implements ModuleCatalog to load MeTTa modules from a file-system directory trying a number of
+/// [FsModuleFormat] formats in succession
+#[derive(Debug)]
+pub struct DirCatalog {
+    path: PathBuf,
+    fmts: Arc<Vec<Box<dyn FsModuleFormat>>>,
+}
+
+impl DirCatalog {
+    /// Internal function to initialize a DirCatalog for a directory in the module search path
+    pub(crate) fn new(path: PathBuf, fmts: Arc<Vec<Box<dyn FsModuleFormat>>>) -> Self {
+        Self {path, fmts}
+    }
+}
+
+impl ModuleCatalog for DirCatalog {
+    fn lookup(&self, name: &str) -> Vec<ModuleDescriptor> {
+
+        //QUESTION: How should we handle modules with an internal "package-name" that differs from their
+        // name in the file system?
+        //
+        //Cargo treats the module's internal "package-name" as authoritative, but that means it's impossible
+        // to "install" a module simply by dropping it into a directory in the search path, because there
+        // needs to be an index of all available modules in that directory.
+        //
+        //For us, I think we want a less formal approach akin to Python's, where we are allowed to drop a
+        // module into a directory, and import it with a naked `import` statement (i.e. no pkg-info entry)
+        // but for that to work, we need to stipulate that it's possible to infer a file name from a
+        // module name.
+        //
+        //NOTE: This is not a limitation across all catalogs, just the `DirCatalog`  If a catalog is able
+        // to maintain its own index of module names, it can store the modules any way it wants to.
+        //
```

Comment:
This also partially becomes a moot point because of the idea that a `ModuleLoader` won't provide a `ModuleDescriptor` anymore.  So there won't be two places that provide names and thus no opportunity for them to conflict.

I'll address this at the same time I implement the fix for https://github.com/trueagi-io/hyperon-experimental/pull/580#discussion_r1491205886